doc/design/block_execution.md

```diff
@@ -0,0 +1,113 @@
+## Overview
+PaddlePaddle represents a neural network as a `Block`. The word block and graph are interchangable in the desgin of PaddlePaddle. A block is like a pair of curly braces in programming languages like C++ and Java, it includes some variables and a sequence of instructions, or, operators. Usually we run a block using a CPU thread. But we want to accelerate the execution of some operators using devices like GPU and FPGA.
+
+There are some challenges here to achieve high performance and scalability when executing a `Block`:
+
+1. Execution Plan. It is often that a kind of device cannot accelerate all operators in a block. This requires an execution plan to use multiple devices. For example, some computational operators run on the GPU and cross-node data communication operators like SendOp and RecvOp on the CPU.
+
+2. Parallel Execution.
+It is often that a computer has more than one acceleration devices. For example, most servers and portable computers like PX2 have more than one CUDA GPUs. We want to make full use of them. Intuitively there are two approaches: (1) running operators in a block on multiple devices, and (2) duplicating the block and running them on multiple devices. The former is also known as model parallelism and the latter data parallelism.
+
+3. Variable Places. Devices often have their onboard memories, and it is usually much more efficient for them to operator variables in the onboard memories. The execution plan should include the placement of variables in a block.
+
+
+Here are the features we want to support:
```

Comment:
Are the below feature sorted by priority? It looks like so because of the "1. 2. 3. ...). But the second point in "3. Model Parallelism" seems very important.

Reply:
It's sorted by priority, data parallelism have higher priority than model parallelism.